### PR TITLE
Add support for Ubuntu Server

### LINKED ISSUE #11: Support for 18.04 Server
Thank you for the great work and it works great for desktop version.

Unfortunately it doesn't work with server edition (ubuntu-18.04.3-live-server-amd64.iso)
/lib/modules is mounted as readonly to the installation media and installation of the zfs kmod failed.

### PR DESCRIPTION
**Pull request by saveriomiroddi**
Ubuntu Server's workflow is very restricted, so the support is quite elaborate.

A graphical glitch appears sometimes during the Ubuntu installer, but it doesn't affect the procedure itself.

Closes #11.

### CODE REVIEW

**Comment by kongkrit** [comment]
Hi saveriomiroddi,

Thanks for the ubuntu server support. I tried to run it, after your script configured & loaded zfs, I was facing a weird bug. When I get to the information screen where I put the username/password, etc. The script would just start installation before I could furnish all the information, and the installation job failed, with some Python stack trace.

I also had to modify your startup a bit to make subiquity run.

I have to do "sudo app install curtin -y" before doing

curl -L https://git.io/JelI5 | sudo bash -

Could it be that curtin actually messed up the installation script? WIthout installaing curtin, subiquity would not start properly.

**Comment by saveriomiroddi** [comment]
> Hi saveriomiroddi,
> 
> Thanks for the ubuntu server support. I tried to run it, after your script configured & loaded zfs, I was facing a weird bug. When I get to the information screen where I put the username/password, etc. The script would just start installation before I could furnish all the information, and the installation job failed, with some Python stack trace.
> 
> I also had to modify your startup a bit to make subiquity run.
> 
> I have to do "sudo app install curtin -y" before doing
> 
> curl -L https://git.io/JelI5 | sudo bash -
> 
> Could it be that curtin actually messed up the installation script? WIthout installaing curtin, subiquity would not start properly.

Can you open two new separate issues for the two problems (curtin install/crash)?

I need to be attached:

- for the curtin problem, a copy of `/tmp/install-zfs.log`, and the error you get if you don't install curtin.
- for the failed job, a copy of `/tmp/install-zfs.log`, and also a copy of Subiquity log, which includes the stack trace; you should find them, in `/var/log/installer/block` (or something similar).

The first is odd, as I can't reproduce it on my system, however, I wonder if it has to do with the path.

Regarding the second, Subiquity installs the system in the background, and, confusingly, when there is an error, it switches window without giving any notice. I was confused as well, at first.

On the issues you'll create, I may ask you further informations.

**Comment by kongkrit** [comment]
Hi saveriomiroddi,

I apologize for speaking too soon. The real issue is HOW I launch:

curl -L https://git.io/JelI5 | sudo bash -

1. If I launch curl from Ctrl+Alt+F2 terminal on hypervisor window, when subiquity starts running (at the screen where you choose the language), the arrow keys stopped working. So, I cannot progress further.

2. If I installed openssh-server and I ssh into the machine and launch your curl, I get the following error after  `install_operating_system_UbuntuServer`. Here's the printout right before it crashes:

```
###############################################################################
# install_operating_system_UbuntuServer
###############################################################################

Parallel unsquashfs: Using 2 processors
6800 inodes (8138 blocks) to write

[=========================================================================================================-] 8138/8138 100%

created 6109 files
created 1198 directories
created 587 symlinks
created 0 devices
created 0 fifos
patching file /tmp/ubiquity_snap_files/lib/python3.6/site-packages/curtin/storage_config.py
patching file /tmp/ubiquity_snap_files/lib/python3.6/site-packages/probert/storage.py
patching file /tmp/ubiquity_snap_files/lib/python3.6/site-packages/curtin/block/__init__.py
patching file /tmp/ubiquity_snap_files/lib/python3.6/site-packages/subiquity/ui/views/installprogress.py
Stopped.
Parallel mksquashfs: Using 2 processors
Creating 4.0 filesystem on /var/lib/snapd/snaps/subiquity_1093.snap, block size 131072.
[=========================================================================================================-] 7551/7551 100%

Exportable Squashfs 4.0 filesystem, gzip compressed, data block size 131072
        compressed data, compressed metadata, compressed fragments, compressed xattrs
        duplicates are removed
Filesystem size 59559.59 Kbytes (58.16 Mbytes)
        22.73% of uncompressed filesystem size (262042.65 Kbytes)
Inode table size 78295 bytes (76.46 Kbytes)
        29.57% of uncompressed inode table size (264799 bytes)
Directory table size 73248 bytes (71.53 Kbytes)
        43.42% of uncompressed directory table size (168693 bytes)
Number of duplicate files found 183
Number of inodes 7894
Number of files 6109
Number of fragments 668
Number of symbolic links  587
Number of device nodes 0
Number of fifo nodes 0
Number of socket nodes 0
Number of directories 1198
Number of ids (unique uids + gids) 1
Number of uids 1
        root (0)
Number of gids 1
        root (0)
Traceback (most recent call last):
  File "/snap/subiquity/1093/usr/bin/subiquity", line 11, in <module>
    load_entry_point('subiquity==0.0.5', 'console_scripts', 'subiquity-tui')()
  File "/snap/subiquity/1093/lib/python3.6/site-packages/subiquity/cmd/tui.py", line 161, in main
    subiquity_interface.run()
  File "/snap/subiquity/1093/lib/python3.6/site-packages/subiquitycore/core.py", line 536, in run
    self.common['loop'].run()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/main_loop.py", line 286, in run
    self._run()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/main_loop.py", line 384, in _run
    self.event_loop.run()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/main_loop.py", line 788, in run
    self._loop()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/main_loop.py", line 825, in _loop
    self._watch_files[fd]()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/raw_display.py", line 404, in <lambda>
    event_loop, callback, self.get_available_raw_input())
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/raw_display.py", line 436, in get_available_raw_input
    codes = self._get_gpm_codes() + self._get_keyboard_codes()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/raw_display.py", line 510, in _get_keyboard_codes
    code = self._getch_nodelay()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/raw_display.py", line 644, in _getch_nodelay
    return self._getch(0)
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/raw_display.py", line 554, in _getch
    return ord(os.read(self._term_input_file.fileno(), 1))
TypeError: ord() expected a character, but string of length 0 found
```

For this second case: here's the tail of `/tmp/install-zfs.log`:

```
Proceed with the configuration as usual, then, at the partitioning stage:

- select `Use an entire disk`
- select `/dev/zd16`
- `Done` -> `Continue` (ignore the warning)
- follow through the installation
- after the security updates are installed, exit to the shell, and follow up with the ZFS installer
' 30 100
+ mount /var/lib/snapd/snaps/subiquity_1093.snap /snap/subiquity/1093
+ PATH=/snap/subiquity/1093/bin:/snap/subiquity/1093/usr/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/s
nap/bin
+ snap run subiquity
```
which actually looks normal.

Tail of `/var/log/installer/subiquity-debug.log` looks like:

```
2019-12-27 09:24:17,360 subiquitycore.prober:67 get_storage: no storage in probe_data, fetching
2019-12-27 09:24:17,361 subiquity.controllers.refresh:86 current version of snap is: '19.07.3'
2019-12-27 09:24:17,361 subiquity.controllers.refresh:122 switching subiquity to stable/ubuntu-18.04.3
2019-12-27 09:24:17,365 subiquitycore.core:538 Exception in controller.run():
Traceback (most recent call last):
  File "/snap/subiquity/1093/lib/python3.6/site-packages/subiquitycore/core.py", line 536, in run
    self.common['loop'].run()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/main_loop.py", line 286, in run
    self._run()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/main_loop.py", line 384, in _run
    self.event_loop.run()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/main_loop.py", line 788, in run
    self._loop()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/main_loop.py", line 825, in _loop
    self._watch_files[fd]()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/raw_display.py", line 404, in <lambda>
    event_loop, callback, self.get_available_raw_input())
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/raw_display.py", line 436, in get_available_raw_input
    codes = self._get_gpm_codes() + self._get_keyboard_codes()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/raw_display.py", line 510, in _get_keyboard_codes
    code = self._getch_nodelay()
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/raw_display.py", line 644, in _getch_nodelay
    return self._getch(0)
  File "/snap/subiquity/1093/usr/lib/python3/dist-packages/urwid/raw_display.py", line 554, in _getch
    return ord(os.read(self._term_input_file.fileno(), 1))
TypeError: ord() expected a character, but string of length 0 found
```

curtin needs not be installed if launched properly through your script.

So I think the cause of it is the unavailability of the arrow keys from the console prevented me from going further.

I apologize again for leading you down the wrong path.

**Comment by saveriomiroddi** [comment]
I've locked the discussion, as the PR itself is completed and merged - the specific terminal issue will be moved to another issue.
